### DISCUSSION TITLE
policy domain="module" meaning

### DISCUSSION THREAD

**Discussion by pgajdos** [ImageMagick/ImageMagick#7350] [General]
I am trying to understand the above element. I tried:

```
<policymap>
  <policy domain="module" rights="read | write" pattern="{PNG}"/>
</policymap>
```
$ magick out.jpg out.png
$ magick out.png out.jpg
$
```
<policymap>
  <policy domain="module" rights="read" pattern="{PNG}"/>
</policymap>
```
$ magick out.jpg out.png
magick: attempt to perform an operation not allowed by the security policy `PNG' @ error/module.c/OpenModule/1268.
$ magick out.png out.jpg
$
```
<policymap>
  <policy domain="module" rights="write" pattern="{PNG}"/>
</policymap>
```
$ magick out.jpg out.png
$ magick out.png out.jpg
$
```
<policymap>
  <policy domain="module" rights="none" pattern="{PNG}"/>
</policymap>
```
$ magick out.jpg out.png
magick: attempt to perform an operation not allowed by the security policy `PNG' @ error/module.c/OpenModule/1268.
$ magick out.png out.jpg
$

When I read in `policy-secure.xml`
```
  <!-- These image types are security risks on read, but write is fine -->
  <policy domain="module" rights="write" pattern="{MSL,MVG,PS,SVG,URL,XPS}"/>
```

I got the impression that `rights="write" pattern="{MSL,MVG,PS,SVG,URL,XPS}` means *allow writing, disallow reading for MSL,MVG,PS,SVG,URL,XPS*. What I am missing?

My intention is something around:
```
<!-- These image types can expose risks on read and write -->
<policy domain="module" rights="none" pattern="{A, B, C}"/>
<!-- These image types are security risks on read, but write is fine -->
<policy domain="module" rights="write" pattern="{D, E, F}"/>
```
i. e. disable some of coders completely, allow writing for other set and allow both reading and writing for the everything else.

**Comment by urban-warrior** [comment]
We cover this use case with the example `config/policy-websafe.xml` configuration file:

```
  <!-- Deny all image modules and specifically exempt reading or writing  web-safe image formats. -->
  <policy domain="module" rights="none" pattern="*" />
  <policy domain="module" rights="read | write" pattern="{BMP,GIF,JPEG,PNG,TIFF,WEBP}"/>
  <policy domain="module" rights="read | write" pattern="{MPC}" stealth="true"/>
  <policy domain="module" rights="write" pattern="{JSON,INFO,PNM,PS}"/>
```

**Comment by pgajdos** [comment]
Thanks @urban-warrior. However, I still do not understand, apologize. To put it another way:

Took `policy-secure.xml` as a base and tried to disallow PNG reading as an example:
```
<!-- These image types are security risks on read, but write is fine -->
<policy domain="module" rights="write" pattern="{MSL,MVG,PS,SVG,URL,XPS,PNG}"/>
```
PNG can be read, still:
```
$ magick logo: out.png
$ magick out.png out.jpg
$ ls -l out.png out.jpg
-rw-r--r-- 1 root root 60304 May 31 12:55 out.jpg
-rw-r--r-- 1 root root 27434 May 31 12:54 out.png
$
```

I also tried to set rights to `none`:
```
<!-- These image types are security risks on read, but write is fine -->
<policy domain="module" rights="none" pattern="{MSL,MVG,PS,SVG,URL,XPS,PNG}"/>
```
Even so magick could read PNG (`out.png` still exists):

```
$ magick logo: out.png
magick: attempt to perform an operation not allowed by the security policy `PNG' @ error/module.c/OpenModule/1268.
$ magick out.png out.jpg
$
```
What am I missing? I have chosen PNG to deviate from `policy-secure.xml` as less as possible, but I get similar result e. g. for PS when I comment `<policy domain="delegate" rights="none" pattern="*"/>` out to be able to use `gs`.
